File: docs/js/mortgage_comparison.js

function init()
{
  var inputs = document.getElementsByTagName("input");
  for(var i=0; i < inputs.length; i++)
  {
    if(inputs[i].type == "number") inputs[i].addEventListener("change", validateNumber);
    inputs[i].addEventListener("change", calcMoney);
  }
  gid("mortgage").addEventListener("change", calcMoney);
  calcMoney();
}


function todays_dollars(monthly, infl, N)
{
  var tmp = 0;
  var out = [];
  for(var i = 0; i < N; i++)
  {
    tmp += monthly / Math.pow(infl, i);
    out.push(tmp);
  }
  return out;
}

function calcMoney()
{
  var N = Number(gid("mortgage").value);
  var p1 = Number(gid("loan1").value);
  var p2 = Number(gid("loan2").value);
  var r1 = Number(gid("rate1").value);
  var r2 = Number(gid("rate2").value);
  var fee1 = Number(gid("fee1").value);
  var fee2 = Number(gid("fee2").value);

  var infl = Number(gid("inflation").value);

  r1 = r1/12/100;
  r2 = r2/12/100;
  infl = Math.pow(1 + infl/100, 1/12);
  var monthly1 = p1*(r1 * Math.pow(1 + r1, N)) / (Math.pow(1 + r1, N) - 1);
  var monthly2 = p2*(r2 * Math.pow(1 + r2, N)) / (Math.pow(1 + r2, N) - 1);

  var pymts1 = todays_dollars(monthly1, infl, N);
  var pymts2 = todays_dollars(monthly2, infl, N);

  var loan2cheaper = (pymts1[0] + fee1 > pymts2[0] + fee2);

  var i = 0;
  for(; i < N; i++)
  {
    if( loan2cheaper && pymts1[i] + fee1 < pymts2[i] + fee2) break;
    if(!loan2cheaper && pymts1[i] + fee1 > pymts2[i] + fee2) break;
  }


  if(loan2cheaper && i == N)
  {
    gid("breakeven").innerHTML = "Loan 2 is always cheaper than loan 1.";
  } else if(loan2cheaper)
  {
    gid("breakeven").innerHTML = "Loan 2 is cheaper for the first " + i + " months.";
  } else if(i == N)
  {
    gid("breakeven").innerHTML = "Loan 1 is always cheaper than loan 2.";
  } else
  {
    gid("breakeven").innerHTML = "Loan 1 is cheaper for the first " + i + " months.";
  }

  gid("monthly1").innerHTML = '$' + monthly1.toFixed(2);
  gid("monthly2").innerHTML = '$' + monthly2.toFixed(2);

  gid("total1").innerHTML = '$' + pymts1[N-1].toFixed(2);
  gid("total2").innerHTML = '$' + pymts2[N-1].toFixed(2);

}

File: R/mortgage_comparison.R


html(
  class = "finance theme-bg",
  HTMLhead(
    titl = "Mortgage Comparison", js = "mortgage_comparison",
    desc = "A comparison of two mortgages",
    date = "2019-07-30"
  ),
  body(
    navbar(),
    disclaimer(
      "This app is only intended to be used as a guide to help decide ",
      "which mortgage to select. It is a simplification, and as such, actual costs and savings ",
      "may differ from what is presented here."
    ),
    div(
      class="row",
      div(
        class="col width-3 inputs-panel bordered",
        h2("Input"),
        label(`for`="mortgage", "Mortgage"),
        select(
          id="mortgage",
          option(value="180", "15 year fixed"),
          option(value="360", "30 year fixed", selected = "selected")
        ),
        label(`for`="inflation", "Inflation Rate (%)"),
        input(type="number", id="inflation", value=2, step=1, min=0)
      ),
      div(
        class = "col width-9 outputs-panel",
        tags$table(
          thead(tr(th(), th("Loan 1"), th("Loan 2"))),
          tbody(
            class="inputs-table thickbottom",
            tr(
              th("Loan amount"),
              td(input(type="number", id="loan1", step=1000, min=0, value=250000)),
              td(input(type="number", id="loan2", step=1000, min=0, value=250000))
            ),
            tr(
              th("Rate (%)"),
              td(input(type="number", id="rate1", value=3.750, step=0.125, min=0, max=50)),
              td(input(type="number", id="rate2", value=3.875, step=0.125, min=0, max=50))
            ),
            tr(
              th("Fees or credits to get this loan (e.g., points)"),
              td(input(type="number", id="fee1", step=100, value=1000)),
              td(input(type="number", id="fee2", step=100, value=0))
            )
          ),
          tbody(
            class="outputs-table",
            tr(
              th("Total paid (in today's dollars)"),
              td(id="total1"),
              td(id="total2")
            ),
            tr(
              th("Monthly mortgage payments"),
              td(id="monthly1"),
              td(id="monthly2")
            ),
            tr(
              th("Which is better?"),
              td(id="breakeven", colspan=2)
            )
          )
        )
      )
    )
  )
) %>%
  write2file(file = "docs/finance/mortgage_comparison.html")
